packages/integration_test/lib/integration_test.dart

```diff
@@ -195,28 +195,29 @@ class IntegrationTestWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding
     results[description] ??= _success;
   }
 
-  vm.VmService _vmService;
+  vm.VmService? _cachedVmService;
+  Future<vm.VmService> get _vmService async {
+    if (_cachedVmService == null) {
+      final developer.ServiceProtocolInfo info =
+          await developer.Service.getInfo();
+      assert(info.serverUri != null);
+      _cachedVmService = await vm_io.vmServiceConnectUri(
+        'ws://localhost:${info.serverUri!.port}${info.serverUri!.path}ws',
+      );
+    }
+    return _cachedVmService!;
+  }
 
   /// Initialize the [vm.VmService] settings for the timeline.
   @visibleForTesting
   Future<void> enableTimeline({
     List<String> streams = const <String>['all'],
-    @visibleForTesting vm.VmService vmService,
+    @visibleForTesting vm.VmService? vmService,
   }) async {
-    assert(streams != null);
+    assert(streams != null); // ignore: unnecessary_null_comparison
     assert(streams.isNotEmpty);
-    if (vmService != null) {
-      _vmService = vmService;
-    }
-    if (_vmService == null) {
-      final developer.ServiceProtocolInfo info =
-          await developer.Service.getInfo();
-      assert(info.serverUri != null);
-      _vmService = await vm_io.vmServiceConnectUri(
-        'ws://localhost:${info.serverUri.port}${info.serverUri.path}ws',
-      );
-    }
-    await _vmService.setVMTimelineFlags(streams);
+    vmService ??= await _vmService;
```

Comment:
I think this should be `if (vmService != null) _cachedVmService = vmService`, so that `traceTimeline` will use the provided `vmService`.

Reply:
I've made the private member a future. I'm also not clear on this, I assumed the param was only for testing.

Reply:
But this parameter is intended to replace the standard `vmService` for testing purposes. If I understand it correctly, your version `traceTimeline` will still call the `_cachedVmService` obtained by `vm_io.vmServiceConnectUri`, but what it should call is the `vmService` provided here.

Reply:
Hmm.  We should add it as a param there then or move it to the ctor/ensureInitialized  It's pretty strange to have a method alter a private member for tests.

Reply:
I've cleaned this up a bit, let me know what you think.